source/common/tracing/opentracing_driver_impl.cc

```diff
@@ -0,0 +1,168 @@
+      return opentracing::make_unexpected(opentracing::lookup_key_not_supported_error);
+    default:
+      NOT_REACHED;
+    }
+  }
+
+  opentracing::expected<void> ForeachKey(OpenTracingCb f) const override {
+    request_headers_.iterate(header_map_callback, static_cast<void*>(&f));
+    return {};
+  }
+
+private:
+  const Http::HeaderMap& request_headers_;
+
+  static Http::HeaderMap::Iterate header_map_callback(const Http::HeaderEntry& header,
+                                                      void* context) {
+    OpenTracingCb* callback = static_cast<OpenTracingCb*>(context);
+    opentracing::string_view key{header.key().c_str(), header.key().size()};
+    opentracing::string_view value{header.value().c_str(), header.value().size()};
+    if ((*callback)(key, value)) {
+      return Http::HeaderMap::Iterate::Continue;
+    } else {
+      return Http::HeaderMap::Iterate::Break;
+    }
+  }
+};
+} // namespace
+
+OpenTracingSpan::OpenTracingSpan(bool use_single_header_propagation, bool use_tracer_propagation,
+                                 std::unique_ptr<opentracing::Span>&& span)
+    : use_single_header_propagation_(use_single_header_propagation),
+      use_tracer_propagation_(use_tracer_propagation), span_(std::move(span)) {}
+
+void OpenTracingSpan::finishSpan() { span_->Finish(); }
+
+void OpenTracingSpan::setOperation(const std::string& operation) {
+  span_->SetOperationName(operation);
+}
+
+void OpenTracingSpan::setTag(const std::string& name, const std::string& value) {
+  span_->SetTag(name, value);
+}
+
+void OpenTracingSpan::injectContext(Http::HeaderMap& request_headers) {
+  if (use_single_header_propagation_) {
+    // Inject the span context using Envoy's single-header format.
+    std::ostringstream oss;
+    const opentracing::expected<void> was_successful =
+        span_->tracer().Inject(span_->context(), oss);
+    if (!was_successful) {
+      ENVOY_LOG(warn, "Failed to inject span context: {}", was_successful.error().message());
+      return;
+    }
+    const std::string current_span_context = oss.str();
+    request_headers.insertOtSpanContext().value(
+        Base64::encode(current_span_context.c_str(), current_span_context.length()));
+  }
+
+  if (use_tracer_propagation_) {
+    // Inject the context using the tracer's standard HTTP header format.
+    const OpenTracingHTTPHeadersWriter writer{request_headers};
+    const opentracing::expected<void> was_successful =
+        span_->tracer().Inject(span_->context(), writer);
+    if (!was_successful) {
+      ENVOY_LOG(warn, "Failed to inject span context: {}", was_successful.error().message());
+      return;
+    }
+  }
+}
+
+SpanPtr OpenTracingSpan::spawnChild(const Config&, const std::string& name, SystemTime start_time) {
+  std::unique_ptr<opentracing::Span> ot_span = span_->tracer().StartSpan(
+      name, {opentracing::ChildOf(&span_->context()), opentracing::StartTimestamp(start_time)});
+  if (ot_span == nullptr) {
```

Comment:
when does this return nullptr?

Reply:
Most of the OpenTracing API is noexcept. LightStep's StartSpan would only return nullptr if something like `std::bad_alloc` was thrown while it was trying to allocate memory. I could change this to use RELEASE_ASSERT since nullptr would likely mean the application is on the verge of crashing.

Reply:
Yeah preference to either use RELEASE_ASSERT for these types of things or in the case of nullptr just let Envoy crash via null-ptr dereference.